src/logging/mod.rs

```diff
@@ -0,0 +1,104 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{io, ops::Deref};
+
+use failure::err_msg;
+use rocket::{Request, State};
+use slog::{self, Drain, Record, Serializer, KV};
+use slog_async;
+use slog_mozlog_json::MozLogJson;
+use slog_term;
+
+use app_errors::Result;
+use settings::*;
+
+lazy_static! {
+    static ref LOGGER_NAME: String =
+        format!("{}-{}", env!("CARGO_PKG_NAME"), env!("CARGO_PKG_VERSION"));
+    static ref MSG_TYPE: String = format!("{}:log", env!("CARGO_PKG_NAME"));
+}
+
+#[derive(Clone)]
+struct RequestMozlogFields {
+    method: &'static str,
+    path: String,
+    remote: Option<String>,
+    agent: Option<String>,
+}
+
+impl RequestMozlogFields {
+    pub fn from_request(request: &Request) -> RequestMozlogFields {
+        let headers = request.headers();
+        RequestMozlogFields {
+            method: request.method().as_str(),
+            path: request.uri().to_string(),
+            agent: headers.get_one("User-Agent").map(&str::to_owned),
+            remote: headers
+                .get_one("X-Forwarded-For")
+                .map(&str::to_owned)
+                .or(request.remote().map(|addr| addr.ip().to_string())),
+        }
+    }
+}
+
+impl KV for RequestMozlogFields {
+    fn serialize(&self, _: &Record, serializer: &mut Serializer) -> slog::Result {
+        if let Some(ref agent) = self.agent {
+            serializer.emit_str("agent", agent)?;
+        }
+        if let Some(ref remote) = self.remote {
+            serializer.emit_str("remoteAddressChain", remote)?;
+        }
+        serializer.emit_str("path", &self.path)?;
+        serializer.emit_str("method", self.method)?;
+
+        Ok(())
+    }
+}
+
+pub struct MozlogLogger(slog::Logger);
+
+impl MozlogLogger {
+    pub fn new(settings: &Settings) -> Result<MozlogLogger> {
+        let json_logging = match settings.mozlog {
+            Some(json_logging) => json_logging,
+            None => Err(err_msg("Invalid FXA_EMAIL_MOZLOG"))?,
```

Comment:
Rather than do this, let's just remove the `Option` part from the setting so that we can guarantee it's always there. Basically stuff that is already in `config/default.json`, there's no need for it to be optional because of the way options stack.